collector/os_release.go

```diff
@@ -0,0 +1,179 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	envparse "github.com/hashicorp/go-envparse"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	etcOSRelease    = "/etc/os-release"
+	usrLibOSRelease = "/usr/lib/os-release"
+	versionRegex    = regexp.MustCompile(`^[0-9]+\.?[0-9]*`)
+)
+
+type osRelease struct {
+	Name            string
+	ID              string
+	IDLike          string
+	PrettyName      string
+	Variant         string
+	VariantID       string
+	Version         string
+	VersionID       string
+	VersionCodename string
+	BuildID         string
+	ImageID         string
+	ImageVersion    string
+}
+
+type osReleaseCollector struct {
+	infoDesc             *prometheus.Desc
+	logger               log.Logger
+	mtimeEtcOSRelease    time.Time
+	mtimeUsrLibOSRelease time.Time
+	os                   *osRelease
+	osMutex              sync.Mutex
+	version              float64
+	versionDesc          *prometheus.Desc
+}
+
+func init() {
+	registerCollector("os", defaultEnabled, NewOSCollector)
+}
+
+// NewOSCollector returns a new Collector exposing os-release information.
+func NewOSCollector(logger log.Logger) (Collector, error) {
+	return &osReleaseCollector{
+		logger: logger,
+		infoDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "info"),
+			"A metric with a constant '1' value labeled by build_id, id, id_like, image_id, image_version, "+
+				"name, pretty_name, variant, variant_id, version, version_codename, version_id.",
+			[]string{"build_id", "id", "id_like", "image_id", "image_version", "name", "pretty_name",
+				"variant", "variant_id", "version", "version_codename", "version_id"}, nil,
+		),
+		versionDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "version"),
+			"Metric containing the major.minor part of the OS version.",
+			nil, nil,
+		),
+	}, nil
+}
+
+func parseOSRelease(r io.Reader) (*osRelease, error) {
+	env, err := envparse.Parse(r)
+	return &osRelease{
+		Name:            env["NAME"],
+		ID:              env["ID"],
+		IDLike:          env["ID_LIKE"],
+		PrettyName:      env["PRETTY_NAME"],
+		Variant:         env["VARIANT"],
+		VariantID:       env["VARIANT_ID"],
+		Version:         env["VERSION"],
+		VersionID:       env["VERSION_ID"],
+		VersionCodename: env["VERSION_CODENAME"],
+		BuildID:         env["BUILD_ID"],
+		ImageID:         env["IMAGE_ID"],
+		ImageVersion:    env["IMAGE_VERSION"],
+	}, err
+}
+
+func (c *osReleaseCollector) updateIfChanged(f *os.File, mtime *time.Time) error {
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	t := stat.ModTime()
+	if t != *mtime {
```

Comment:
Lets do
```
if t == *mTime {
   return nil
}
```

Early returns are go best practice, see https://golang.org/doc/effective_go#if

Reply:
Changed. The code looks better that way.